cloudinit/cmd/clean.py

```diff
@@ -25,6 +25,14 @@
 )
 
 ETC_MACHINE_ID = "/etc/machine-id"
+GEN_NET_CONFIG_FILES = [
+    "/etc/NetworkManager/conf.d/99-cloud-init.conf",
+    "/etc/NetworkManager/conf.d/30-cloud-init-ip6-addr-gen-mode.conf",
+    "/etc/network/interfaces.d/50-cloud-init.cfg",
```

Comment:
Sounds good to me.

Reply:
But I think the above enhacement could also be a separate patch.

Reply:
```suggestion
    +RENDERER_NET_CONFIG_FILES = {
+    "eni": [
+        ("/etc/network/interfaces.d/50-cloud-init.cfg", None),
+        ("/etc/udev/rules.d/70-persistent-net.rules", None),
+    ],
+    "network_manager": [
+        (
+            "/etc/NetworkManager/system-connections/cloud-init-*.nmconnection",
+            None,
+        ),
+        (
+            "/etc/NetworkManager/conf.d/30-cloud-init-ip6-addr-gen-mode.conf",
+            None,
+        ),
+    ],
+    "netplan": [
+        ("/etc/netplan/50-cloud-init.yaml", None),
+    ],
+    "networkd": [
+        ("/etc/systemd/network/10-cloud-init-*.network"),
+    ],
+    "sysconfig": [
+        ("/etc/udev/rules.d/70-persistent-net.rules", None),
+        ("/etc/NetworkManager/conf.d/99-cloud-init.conf", None),
+    ],
+}
```
I think this is a wrong direction anyway. Instead of keeping the information here, the respective net renderers should publish the files they are generating to a common place. That way, there would be no need to track it like this here per renderer.

Reply:
works for me for this iteration, I was thinking the dict would lay groundwork for per-Renderer `config_files` method for each renderer in the future. But, I agree it's probably overkill for this pass and probably adds complexity that we'll replace anyway.

Aren't we still missing the file path glob and removal of the following file paths?
- `"/etc/systemd/network/10-cloud-init-*.network"`  (written by cloudinit.net.networkd.Renderer)
- `"/etc/NetworkManager/system-connections/cloud-init-*.nmconnection"` (written by network_manager.Renderer)


I think we just need to track/remove those paths and use glob.glob around any GEN_NET_CONFIG_FILES and we get coverage for those files which will have an interface enecoded  into the cloud-init network config path name.


Here's a diff which adds the two additional GET_NET_CONFIG_FILES paths, uses glob.glob to match  and fixes unit tests. 

Additionally I think we should probably be covering cloud-init clean with its own integration test at tests/integration_tests/cmd/test_clean.py  I've added that below as well


```diff
diff --git a/cloudinit/cmd/clean.py b/cloudinit/cmd/clean.py
index 0f344e3a7..f51788ed2 100755
--- a/cloudinit/cmd/clean.py
+++ b/cloudinit/cmd/clean.py
@@ -28,6 +28,8 @@ ETC_MACHINE_ID = "/etc/machine-id"
 GEN_NET_CONFIG_FILES = [
     "/etc/NetworkManager/conf.d/99-cloud-init.conf",
     "/etc/NetworkManager/conf.d/30-cloud-init-ip6-addr-gen-mode.conf",
+    "/etc/NetworkManager/system-connections/cloud-init-*.nmconnection",
+    "/etc/systemd/network/10-cloud-init-*.network",
     "/etc/network/interfaces.d/50-cloud-init.cfg",
 ]
 GEN_SSH_CONFIG_FILES = [
@@ -119,8 +121,9 @@ def remove_artifacts(remove_logs, remove_seed=False, remove_config=None):
         for log_file in get_config_logfiles(init.cfg):
             del_file(log_file)
     if remove_config and set(remove_config).intersection(["all", "network"]):
-        for conf in GEN_NET_CONFIG_FILES:
-            del_file(conf)
+        for path in GEN_NET_CONFIG_FILES:
+            for conf in glob.glob(path):
+                del_file(conf)
     if remove_config and set(remove_config).intersection(
         ["all", "ssh_config"]
     ):
diff --git a/tests/integration_tests/cmd/test_clean.py b/tests/integration_tests/cmd/test_clean.py
new file mode 100644
index 000000000..37df947b1
--- /dev/null
+++ b/tests/integration_tests/cmd/test_clean.py
@@ -0,0 +1,86 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+import re
+
+import pytest
+
+from tests.integration_tests.instances import IntegrationInstance
+
+USER_DATA = """\
+#cloud-config
+write_files:
+- path: /etc/network/interfaces.d/50-cloud-init.cfg
+  content: cloud-init test was here
+  permissions: '0644'
+  owner: root:root
+- path: /etc/cloud/clean.d/runme.sh
+  encoding: b64
+  content: IyEvYmluL3NoCmVjaG8gIi9ldGMvY2xvdWQvY2xlYW4uZC9ydW5tZS5zaCBSQU4iCg==
+  permissions: '0755'
+  owner: root:root
+- path: /etc/systemd/network/10-cloud-init-eth0.network
+  content: cloud-init test was here
+  permissions: '0644'
+  owner: root:root
+- path: /etc/cloud/clean.d/dontrunme.sh
+  content: '#!/bin/sh\necho DID NOT RUN BECAUSE NO EXEC PERMS'
+  permissions: '0644'
+  owner: root:root
+"""
+
+
+@pytest.mark.user_data(USER_DATA)
+class TestCleanCommand:
+    def test_clean_by_param(self, class_client: IntegrationInstance):
+        """Clean with various params alters expected files without error"""
+        assert class_client.execute("cloud-init status --wait").ok
+        result = class_client.execute("cloud-init clean")
+        assert (
+            result.ok
+        ), "non-zero exit on cloud-init clean runparts of /etc/cloud/clean.d"
+        # Log files are not removed without --logs
+        log_paths = (
+            "/var/log/cloud-init.log",
+            "/var/log/cloud-init-output.log",
+        )
+        net_cfg_paths = (
+            "/etc/network/interfaces.d/50-cloud-init.cfg",
+            "/etc/netplan/50-cloud-init.yaml",
+            "/etc/systemd/network/10-cloud-init-eth0.network",
+        )
+        for path in log_paths + net_cfg_paths:
+            assert class_client.execute(
+                f"test -f {path}"
+            ).ok, f"Missing expected file {path}"
+        # /etc/cloud/clean.d runparts scripts are run if executable
+        assert result.stdout == "/etc/cloud/clean.d/runme.sh RAN"
+
+        # Log files removed with --logs
+        assert class_client.execute("cloud-init clean --logs").ok
+        for path in log_paths:
+            assert class_client.execute(
+                f"test -f {path}"
+            ).failed, f"Unexpected file found {path}"
+        for path in net_cfg_paths:
+            assert class_client.execute(
+                f"test -f {path}"
+            ).ok, f"Missing expected file {path}"
+
+        prev_machine_id = class_client.read_from_file("/etc/machine-id")
+        assert re.match(
+            r"^[a-f0-9]{32}$", prev_machine_id
+        ), f"Unexpected machine-id format {prev_machine_id}"
+
+        # --machine-id sets /etc/machine-id
+        assert class_client.execute("cloud-init clean --machine-id").ok
+        machine_id = class_client.read_from_file("/etc/machine-id")
+        assert machine_id != prev_machine_id
+        assert "uninitialized" == machine_id
+
+        # --configs remove network scope
+        assert class_client.execute("cloud-init clean --configs network").ok
+
+        import pdb; pdb.set_trace()
+        for path in log_paths + net_cfg_paths:
+            assert class_client.execute(
+                f"test -f {path}"
+            ).failed, f"Unexpected file found {path}"
diff --git a/tests/unittests/cmd/test_clean.py b/tests/unittests/cmd/test_clean.py
index 862a7d47f..eaad142ec 100644
--- a/tests/unittests/cmd/test_clean.py
+++ b/tests/unittests/cmd/test_clean.py
@@ -7,7 +7,7 @@ import pytest
 
 import cloudinit.settings
 from cloudinit.cmd import clean
-from cloudinit.util import ensure_dir, sym_link
+from cloudinit.util import ensure_dir, sym_link, write_file
 from tests.unittests.helpers import mock, wrap_and_call
 
 MyPaths = namedtuple("MyPaths", "cloud_dir")
@@ -80,11 +80,18 @@ class TestClean:
         for conf_path in TEST_GEN_NET_CONFIG_FILES:
             assert conf_path.exists() is False, f"Unexpected {conf_path}"
             ensure_dir(os.path.dirname(conf_path))
-            conf_path.write(f"#generated by cloud-init\ntouch {conf_path}\n")
+            if "*" in conf_path.strpath:
+                # Expand glob path to interface-specific eth0 filename
+                eth0_path = conf_path.strpath.replace("*", "eth0")
+                write_file(
+                    eth0_path, f"#generated by cloud-init test {eth0_path}"
+                )
+            else:
+                conf_path.write(f"#generated by cloud-init test {conf_path}")
 
         with mock.patch(
             "cloudinit.cmd.clean.GEN_NET_CONFIG_FILES",
-            TEST_GEN_NET_CONFIG_FILES,
+            [f.strpath for f in TEST_GEN_NET_CONFIG_FILES],
         ):
             retcode = wrap_and_call(
                 "cloudinit.cmd.clean",
@@ -95,7 +102,16 @@ class TestClean:
             )
 
         for conf_path in TEST_GEN_NET_CONFIG_FILES:
-            assert conf_path.exists() is False, f"Unexpected file {conf_path}"
+            if "*" in conf_path.strpath:
+                # Expand glob path to interface-specific eth0 filename
+                assert (
+                    os.path.exists(conf_path.strpath.replace("*", "eth0"))
+                    is False
+                ), f"Unexpected file {conf_path.strpath.replace('*', 'eth0')}"
+            else:
+                assert (
+                    conf_path.exists() is False
+                ), f"Unexpected file {conf_path}"
         assert 0 == retcode
 
     def test_remove_ssh_conf(self, clean_paths, init_class):
@@ -149,7 +165,7 @@ class TestClean:
 
         with mock.patch(
             "cloudinit.cmd.clean.GEN_NET_CONFIG_FILES",
-            TEST_GEN_NET_CONFIG_FILES,
+            [f.strpath for f in TEST_GEN_NET_CONFIG_FILES],
         ), mock.patch(
             "cloudinit.cmd.clean.GEN_SSH_CONFIG_FILES",
             TEST_GEN_SSH_CONFIG_FILES,
```

Reply:
applied.